src/Common/src/TypeSystem/IL/Stubs/TypeGetTypeMethodThunk.cs

```diff
@@ -0,0 +1,178 @@
+
+using Internal.TypeSystem;
+
+using Debug = System.Diagnostics.Debug;
+
+namespace Internal.IL.Stubs
+{
+    /// <summary>
+    /// Helps model the behavior of calls to various Type.GetType overloads. Type.GetType is required to search
+    /// the calling assembly for a matching type if the type name supplied by the user code was not assembly qualified.
+    /// This thunk calls a helper method, passing it a string for what should be considered the "calling assembly".
+    /// </summary>
+    internal class TypeGetTypeMethodThunk : ILStubMethod
+    {
+        private readonly MethodDesc _helperMethod;
+
+        public TypeGetTypeMethodThunk(TypeDesc owningType, MethodSignature signature, MethodDesc helperMethod, string defaultAssemblyName)
+        {
+            OwningType = owningType;
+            Signature = signature;
+            DefaultAssemblyName = defaultAssemblyName;
+
+            _helperMethod = helperMethod;
+        }
+
+        public override TypeSystemContext Context
+        {
+            get
+            {
+                return OwningType.Context;
+            }
+        }
+
+        public override string Name
+        {
+            get
+            {
+                return $"{_helperMethod.Name}_{Signature.Length}_{DefaultAssemblyName}";
+            }
+        }
+
+        public override TypeDesc OwningType
+        {
+            get;
+        }
+
+        public override MethodSignature Signature
+        {
+            get;
+        }
+
+        public string DefaultAssemblyName
+        {
+            get;
+        }
+
+        public override MethodIL EmitIL()
+        {
+            ILEmitter emit = new ILEmitter();
+            ILCodeStream codeStream = emit.NewCodeStream();
+
+            Debug.Assert(Signature[0] == _helperMethod.Signature[0]);
+            codeStream.EmitLdArg(0);
+
+            Debug.Assert(_helperMethod.Signature[1].IsString);
+            codeStream.Emit(ILOpcode.ldstr, emit.NewToken(DefaultAssemblyName));
+
+            for (int i = 2; i < _helperMethod.Signature.Length; i++)
+            {
+                // The helper method could be expecting more arguments than what we have - check for that
+                // The thunk represents one of the 6 possible overloads:
+                // (String), (String, bool), (String, bool, bool)
+                // (String, Func<...>, Func<...>), (String, Func<...>, Func<...>, bool), (String, Func<...>, Func<...>, bool, bool)
+                // We only need 2 helpers to support all 6 overloads. The default value for the bools is false.
+
+                if (i - 1 < Signature.Length)
+                {
+                    // Pass user's parameter
+                    Debug.Assert(_helperMethod.Signature[i] == Signature[i - 1]);
+                    codeStream.EmitLdArg(i - 1);
+                }
+                else
+                {
+                    // Pass a default value
```

Comment:
> And also not call it GetType transform, but something more general

This will need some additional thoughts on generating a unique name for the generated helper methods, etc. Let's cross that bridge when we get there. I'm actually leaning towards just making a separate thunk class for `GetExecutingAssembly` too.

> It may look better to have the helpers with the hidden argument on the same type that they are supporting

I feel the opposite way on this: compiler helpers should be in `Internal.Runtime.CompilerHelpers`, not on the type they support. You could make the same argument for e.g. `LdTokenHelpers` and put `GetRuntimeTypeHandle` on `System.RuntimeTypeHandle`, or put a `Throw()` method on all exception types instead of having a `ThrowHelpers` class but I don't think it helps in keeping the class library code tidy.